### DISCUSSION TITLE
With MSVC build tools setup, how to have the `*.obj` and `.pdb` files be generated in separated directory?

### DISCUSSION THREAD

**Discussion by aenlemmea** [microsoft/vscode-cpptools#11876] [Q&A] [answered]
Currently they generate under workspace folder. The built binary successfully builds in `build` folder with my `tasks.json` but how do I achieve the same for the obj and .pdb files. I tried doing that with `/Fo` and `/Fd` but it does not work.

```
{
    "tasks": [
        {
            "type": "cppbuild",
            "label": "C/C++: cl.exe build active file",
            "command": "cl.exe",
            "args": [
                "/Zi",
                "/EHsc",
                "/sdl",
                "/nologo",
                "/Fe${fileDirname}\\build\\${fileBasenameNoExtension}.exe",
                "/Fd${workspaceFolder}\\obj\\vc140.pdb",
                "/Fo${workspaceFolder}\\obj\\${fileBasenameNoExtension}.obj",
                "${file}"
            ],
            "options": {
                "cwd": "${fileDirname}"
            },
            "problemMatcher": [
                "$msCompile"
            ],
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "detail": "Task generated by Debugger."
        }
    ],
    "version": "2.0.0"
}
```

**Comment by bobbrow** [comment]
Does the task report a failure?  It's possible that the path does not exist and you might need to ensure it exists before trying to compile.

**Reply by aenlemmea** [reply]
Yes it resulted in a failure demanding the use of `/Fs`, however using it again results in the same failure. I tried writing a custom task dependency which would ensure if the folders are there and if not then create them however again lead to the same errors.

So far if I manually create the directories, it works. I just need a automated way to create them as a task, which oddly didn't work out.

**Reply by bobbrow** [reply, marked as answer]
Does this work for you?
```jsonc
    "tasks": [
        {
            "label": "mkdir",
            "type": "shell",
            "command": "if not exist \"${workspaceFolder}\\obj\" md \"${workspaceFolder}\\obj\""
        },
        {
            "type": "cppbuild",
            // ...
            "dependsOn": "mkdir",
        }
    ]

```

**Reply by aenlemmea** [reply]
Yep, worked out. For anyone else running into the issue of powershell being default, you can change it to CMD and this works out.
